applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -488,247 +593,106 @@ int handle_at_gps(enum at_cmd_type cmd_type)
 				return -EINVAL;
 			}
-			err = nrf_modem_gnss_fix_interval_set(interval);
-			if (err) {
-				LOG_ERR("Failed to set fix interval, error: %d", err);
-				return err;
+			if (gnss_agps_enabled && !IS_ENABLED(CONFIG_NRF_CLOUD_AGPS)) {
+				LOG_ERR("Missing A-GPS support (CONFIG_NRF_CLOUD_AGPS).");
+				return -ENOTSUP;
 			}
-			if (at_params_unsigned_short_get(&slm_at_param_list, 3, &timeout) == 0) {
-				err = nrf_modem_gnss_fix_retry_set(timeout);
-				if (err) {
-					LOG_ERR("Failed to set fix retry, error: %d", err);
-					return err;
-				}
-			} /* else leave it to default or previously configured timeout */
-
-			err = gnss_startup(RUN_TYPE_GPS);
-		} else if (op == GPS_STOP && run_type == RUN_TYPE_GPS) {
-			err = gnss_shutdown();
-		} else {
-			err = -EINVAL;
-		} break;
-
-	case AT_CMD_TYPE_READ_COMMAND:
-		rsp_send("\r\n#XGPS: %d,%d\r\n", (int)is_gnss_activated(), run_status);
-		err = 0;
-		break;
-
-	case AT_CMD_TYPE_TEST_COMMAND:
-		rsp_send("\r\n#XGPS: (%d,%d),<interval>,<timeout>\r\n", GPS_STOP, GPS_START);
-		err = 0;
-		break;
-
-	default:
-		break;
-	}
-
-	return err;
-}
-
-#if defined(CONFIG_SLM_NRF_CLOUD) && defined(CONFIG_NRF_CLOUD_AGPS)
-
-static int64_t utc_to_gps_sec(const int64_t utc_sec)
-{
-	return (utc_sec - GPS_TO_UNIX_UTC_OFFSET_SECONDS) + GPS_TO_UTC_LEAP_SECONDS;
-}
-
-static void gps_sec_to_day_time(int64_t gps_sec, uint16_t *gps_day, uint32_t *gps_time_of_day)
-{
-	*gps_day = (uint16_t)(gps_sec / SEC_PER_DAY);
-	*gps_time_of_day = (uint32_t)(gps_sec % SEC_PER_DAY);
-}
-
-/* Handles AT#XAGPS commands. */
-int handle_at_agps(enum at_cmd_type cmd_type)
-{
-	int err = -EINVAL;
-	uint16_t op;
-	uint16_t interval;
-	uint16_t timeout;
-	int64_t utc_now_ms;
 
-	switch (cmd_type) {
-	case AT_CMD_TYPE_SET_COMMAND:
-		err = at_params_unsigned_short_get(&slm_at_param_list, 1, &op);
-		if (err < 0) {
-			return err;
-		}
-		if (op == AGPS_START && slm_nrf_cloud_ready && run_type == RUN_TYPE_NONE) {
-			err = at_params_unsigned_short_get(&slm_at_param_list, 2, &interval);
-			if (err < 0) {
-				return err;
-			}
-			/* GNSS API spec check */
-			if (interval != 0 && interval != 1 &&
-			   (interval < 10 || interval > 65535)) {
+			err = at_params_unsigned_short_get(
+				&slm_at_param_list, 3, &gnss_pgps_enabled);
+			if (err || gnss_pgps_enabled > 1) {
 				return -EINVAL;
 			}
-#if defined(CONFIG_NRF_CLOUD_AGPS_FILTERED)
-			err = nrf_modem_gnss_elevation_threshold_set(
-						CONFIG_NRF_CLOUD_AGPS_ELEVATION_MASK);
-			if (err) {
-				LOG_ERR("Failed to set elevation threshold, error: %d", err);
-				return err;
-			}
-#endif
-			/* no scheduled downloads for peoridical tracking */
-			if (interval >= 10) {
-				err = nrf_modem_gnss_use_case_set(
-						NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START |
-						NRF_MODEM_GNSS_USE_CASE_SCHED_DOWNLOAD_DISABLE);
-				if (err) {
-					LOG_ERR("Failed to set use case, error: %d", err);
-					return err;
-				}
+			if (gnss_pgps_enabled && !IS_ENABLED(CONFIG_NRF_CLOUD_PGPS)) {
```

Comment:
I am wondering whether we need to support the different options in runtime as the PGPS and AGPS availability is decided in the compile time. It would be more straightforward to follow the same compile time flagging and offer only the operation that is flagged in.

I am also not sure if the operation that we have for example, with AGPS + PGPS, but only requested for AGPS is equivalent to compilation with AGPS + requested AGPS. I am sure that SLM codes are equivalent, but in other components there are differences.

Reply:
> By testing and observing the answer is no. Especially in the case you mention. When both A-GPS and P-GPS are compiled in, A-GPS starts relying on P-GPS for processing some of the data. That's why I put some countermeasures in that code, to try to not inject the data from one service (A-GPS/P-GPS) if it is not requested in the AT command. But it may be a foolish endeavor in the end. At least one that is pretty solid and good to keep I think is allowing to not use any of the two services at run time even though one or both have been compiled in. By ignoring events and not triggering any related action it's possible (and the case here) to have the libraries not do anything.

Having countermeasures to hide some of the existing functionality might be problematic in case there are real differences in some scenarios, or if the GNSS libraries change functionality, so that there would be.

Having a possibility to ignore both, would be pretty much similar functionality that we had previously, where you could send GPS, AGPS and PGPS requests. I am pretty certain that neither AGPS or PGPS worked correctly if both were active, so that possibility did not really exist.

Reply:
@VTPeltoketo, @SeppoTakalo, @junqingzou: What are your takes on this?

Reply:
I would say that if you have both AGPS and PGPS compiled, you can activate only both at the same time. Still, I'd like got get opinion from some GNSS-expert.

Reply:
So, according to my understanding we have three options when AGPS and PGPS are activated at compiler time:
- GNSS is disabled
- GNSS is enabled without assistance
- GNSS is enabled with assistance when both AGPS and PGPS are activated